plugin.video.playonbrowser/resources/lib/playon.py

```diff
@@ -0,0 +1,360 @@
+        if URLTYPE == 'upnp': self.chkUPNP()
+            
+            
+    def openURL(self, url):
+        log('openURL, url = ' + str(url))
+        try:
+            cacheResponse = self.cache.get(ADDON_NAME + '.openURL, url = %s'%url)
+            if not cacheResponse:
+                request = urllib2.Request(url)
+                request.add_header('User-Agent','Mozilla/5.0 (Windows; U; MSIE 9.0; Windows NT 9.0; en-US)')
+                cacheResponse = urllib2.urlopen(request, timeout=TIMEOUT).read()
+                self.cache.set(ADDON_NAME + '.openURL, url = %s'%url, cacheResponse, expiration=datetime.timedelta(minutes=5))
+            return cacheResponse
+        except Exception as e: log("openURL Failed! " + str(e), xbmc.LOGERROR)
+        if url == BASE_URL + PLAYON_DATA: self.getIP()
+        
+        
+    def getIP(self):
+        urls  = self.openURL(AUTO_URL).split('|')
+        for url in urls:
+            match = re.findall(r'[0-9]+(?:\.[0-9]+){3}:[0-9]+', url)
+            log('getIP, match = ' + str(match))
+            if len(match) == 0:
+                url = getKeyboard(LANGUAGE(30001),LANGUAGE(30002))
+                if url == False: return
+            if not url.startswith('http'): url = "http://%s"%url
+            BASE_URL = url
+            REAL_SETTINGS.setSetting("playonserver",url)
+            if self.chkIP(url): return
+        
+        
+    def chkIP(self, url=BASE_URL):
+        log('chkIP, url = ' + str(url))
+        results = xmltodict.parse(self.openURL(url + PLAYON_DATA))
+        if results and 'catalog' in results:
+            try:
+                ServerName = results['catalog']['@name']
+                ServerVer = 'PlayOn v.%s'%results['catalog']['@server']
+                ServerMSG = "Connected to [B]%s %s[/B]"%(ServerName,ServerVer)
+                log('chkIP, ServerName = ' + ServerName)
+                log('chkIP, ServerVer = ' + ServerVer)
+                REAL_SETTINGS.setSetting("playonServerid",ServerMSG)
+                xbmcgui.Dialog().notification(ADDON_NAME, LANGUAGE(30011)%ServerName, ICON, 5000)
+                xbmc.executebuiltin("Container.Refresh")
+                return True
+            except Exception as e: log("chkIP Failed! " + str(e), xbmc.LOGERROR)
+        else: xbmcgui.Dialog().notification(ADDON_NAME, LANGUAGE(30010), ICON, 5000)
+        return False
+        
+    
+    def getUPNP(self):
+        log('getUPNP')
+        """ Check Kodi UPNP support. """
+        json_query = ('{"jsonrpc":"2.0","method":"Settings.GetSettingValue","params":{"setting":"services.upnp"},"id":1}')
+        data = json.loads(xbmc.executeJSONRPC(json_query))
+        try:
+            if 'result' in data and 'value' in data['result']: return data['result']['value']
+        except Exception as e: log('getUPNP, Failed! ' + str(e))
+        return False
+        
+    
+    def setUPNP(self):
+        log('setUPNP')
+        """ Enable Kodi UPNP support. """
```

Comment:
I am not the biggest fan of this but I understand upnp needs to be enabled to use the addon. Consider to at least produce some sort of warning before enabling the setting in the future